tools/cldr-code/src/test/java/org/unicode/cldr/api/CldrDataProcessorTest.java

```diff
@@ -0,0 +1,135 @@
+package org.unicode.cldr.api;
+
+import com.ibm.icu.dev.test.TestFmwk;
+import java.util.Arrays;
+import java.util.LinkedHashMap;
+import java.util.Map;
+import java.util.Objects;
+
+import com.google.common.collect.ImmutableMap;
+
+public class CldrDataProcessorTest extends TestFmwk {
+    private static final AttributeKey TERRITORY_TYPE = AttributeKey.keyOf("territory", "type");
+    private static final AttributeKey CURRENCY_TYPE = AttributeKey.keyOf("currency", "type");
+
+    // An overly simplistic value type for currency for testing purposes. In real code you would
+    // probably want an immutable type and a separate builder, or a mutable type just to collect
+    // values that doesn't need equals/hashcode (this class serves 2 purposes in the test).
+    private static final class CurrencyData {
+        final String key;
+        String name = "";
+        String symbol = "";
+
+        CurrencyData(String key) {
+            this.key = key;
+        }
+
+        CurrencyData(String key, String name, String symbol) {
+            this.key = key;
+            this.name = name;
+            this.symbol = symbol;
+        }
+
+        @Override public boolean equals(Object o) {
+            if (o instanceof CurrencyData) {
+                CurrencyData that = (CurrencyData) o;
+                return key.equals(that.key) && name.equals(that.name) && symbol.equals(that.symbol);
+            }
+            return false;
+        }
+
+        @Override public int hashCode() {
+            return Objects.hash(key, name, symbol);
+        }
+
+        @Override public String toString() {
+            return String.format("CurrencyData{name=%s, symbol='%s'}", name, symbol);
+        }
+    }
+
+    // For collecting processed values.
+    private static final class State {
+        ImmutableMap<String, String> names = ImmutableMap.of();
+        ImmutableMap<String, CurrencyData> currencies = ImmutableMap.of();
+
+        void setNames(Map<String, String> map) {
+            names = ImmutableMap.copyOf(map);
+        }
+
+        void setCurrencies(Map<String, CurrencyData> map) {
+            currencies = ImmutableMap.copyOf(map);
+        }
+    }
+
+    private static final CldrDataProcessor<State> VISITOR = createTestVisitor();
+
+    private static CldrDataProcessor<State> createTestVisitor() {
+        // Note that this is deliberately doing things the "messy" way by creating and then copying
+        // a map. This is to show an extra level of processing in tests. You could just have a
+        // value action which adds the territory to a map in the State object.
+        CldrDataProcessor.Builder<State> builder = CldrDataProcessor.builder();
+        builder
+            .addAction(
+                "//ldml/localeDisplayNames/territories",
+                () -> new LinkedHashMap<String, String>(),
```

Comment:
maybe some explanation here: 

// "In this example, the subprocessor action operates on territories. The subprocessor state* is the Map<String code,String displayName> which collects the territory translations. When all of the translations are done, setNames() collects the subprocessor state into the main state

(* or context etc)